criu/mem.c

```diff
@@ -131,6 +140,171 @@ int init_pidfd_store_sk(pid_t pid, int sk)
 	return -1;
 }
 
+void free_pidfd_store(void)
+{
+	int i;
+	struct pidfd_entry *entry;
+	struct hlist_node *tmp;
+
+	for (i = 0; i < PIDFD_HASH_SIZE; i++) {
+		hlist_for_each_entry_safe(entry, tmp, &pidfd_hash[i], hash) {
+			close(entry->pidfd);
+			xfree(entry);
+		}
+	}
+
+	close_safe(&pidfd_store_sk);
+}
+
+int init_pidfd_store_hash(void)
+{
+	int i, cnt, ret;
+	struct pidfd_entry *entry;
+
+	for (i = 0; i < PIDFD_HASH_SIZE; i++)
+		INIT_HLIST_HEAD(&pidfd_hash[i]);
+
+	/* Skip building pidfd_hash if pidfd_store_sk is not initialized */
+	if (pidfd_store_sk == -1)
+		return 0;
+
+	/*
+	 * Drain all queued pidfd entries in pidfd_store_sk from
+	 * the last predump into pidfd_hash.
+	 */
+	cnt = 0;
+	while (1) {
+		entry = xmalloc(sizeof(struct pidfd_entry));
+		INIT_HLIST_NODE(&entry->hash);
+
+		ret = __recv_fds(pidfd_store_sk, &entry->pidfd, 1, &entry->pid, sizeof(pid_t), MSG_DONTWAIT);
+		if (ret == -EAGAIN || ret == -EWOULDBLOCK) {
+			/* No more fds to read */
+			xfree(entry);
+			goto check_empty;
+		} else if (ret) {
+			pr_perror("Can't read pidfd");
+			xfree(entry);
+			free_pidfd_store();
+			return ret;
+		}
+
+		cnt++;
+		hlist_add_head(&entry->hash, &pidfd_hash[entry->pid % PIDFD_HASH_SIZE]);
+	}
+
+check_empty:
+	/*
+	 * If no pidfds exist in pidfd_store. This would cause full page
+	 * dumps which goes against the purpose of the pidfd store.
+	 * This is probably due to sending a different pidfd_store socket.
+	 */
+	if (cnt == 0 && opts.img_parent) {
+		pr_err("No pidfds found in pidfd store\n");
+		pr_err("The same socket from the previous iteration should be passed\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+static struct pidfd_entry *find_pidfd_entry_by_pid(pid_t pid)
+{
+	struct pidfd_entry *entry;
+	struct hlist_head *chain;
+
+	chain = &pidfd_hash[pid % PIDFD_HASH_SIZE];
+	hlist_for_each_entry(entry, chain, hash) {
+		if (entry->pid == pid)
+			return entry;
+	}
+
+	return NULL;
+}
+
+/*
+ * 1 - task closed
+ * 0 - task still running
+ * -1 - error
+ */
+static int check_pidfd_entry_state(struct pidfd_entry *entry)
+{
+	struct pollfd pollfd;
+	int ret, restart_cnt = 0;
+	const int MAX_RESTARTS = 10; /* Reasonable limit to avoid getting stuck */
+
+	/*
+	 * When there is data to read from the pidfd, it means
+	 * that the task associated with this pidfd is closed.
+	 */
+	pollfd.fd = entry->pidfd;
+	pollfd.events = POLLIN;
+
+	while (1) {
+		ret = poll(&pollfd, 1, 0);
```

Comment:
poll with the zero timeout can't return EINTR, can it?

Reply:
@avagin I encountered this behavior before once when polling with zero timeout in Golang, This was most likely due to the Go runtime bombarding the thread with ```SIGURG```. I am not sure if there is a scenario where CRIU would receive a high frequency of signals that would cause zero timeout poll to interrupt, so I went with the safe route.

Your perspective on this would be highly valuble.

Reply:
@avagin Looks like it actually can return EINTR.

```
do_poll()
{
...
                if (!count) {
                        count = wait->error;
                        if (signal_pending(current))
                                count = -ERESTARTNOHAND;
                }
}
```

If signal_pending is true. I don't see any explicit protection from it.

Reply:
Ok, I had to be more detailed. I actually mean that it will not return EINTR if there is any event.

Reply:
@ZeyadYasser Anyway, the code in check_pidfd_entry_state() is badly broken. Shame on me that I initially missed it...

!!!!One should never rely on errno if libc syscall is successful!!!!

Imagine on the first iteration you get no event and a signal and thus get errno == EINTR or errno was already EINTR before we even started poll-ing because of previous calls, then all later iterations would be successful but errno would still continue to be EINTR, and we would cycle 10 times for no reason.

Also it is known practice for glibc to set errno to some error even in case of successful glibc syscall, because it can do "retries" inside glibc and first try may set errno, while second try would be successful.

Probably it was the case with Golang too, I've seen places in docker code where they forget to check retcode and wrongly rely on errno =)

So we should probably remove the while loop.

Reply:
My bad, I didn't know that a call could be successful and errno not changing.
Thank you